Source/Diagnostics/ReducedDiags/LoadBalanceCosts.cpp

```diff
@@ -0,0 +1,192 @@
+/* Copyright 2019-2020 Michael Rowan, Yinjian Zhao
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "LoadBalanceCosts.H"
+#include "WarpX.H"
+
+
+using namespace amrex;
+
+// constructor
+LoadBalanceCosts::LoadBalanceCosts (std::string rd_name)
+    : ReducedDiags{rd_name}
+{
+
+}
+
+// function that gathers costs
+void LoadBalanceCosts::ComputeDiags (int step)
+{
+    // get WarpX class object
+    auto& warpx = WarpX::GetInstance();
+
+    // for now, costs reduced diagnostic only works with `costs_heuristic`,
+    // but this will work for timer based costs as well after changing from
+    // multifab to vector
+    const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(0);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(cost_heuristic != nullptr,
+        "LoadBalanceCosts reduced diagnostic works only with heuristic costs update.");
+
+    // judge if the diags should be done
+    // costs is initialized only if we're doing load balance
+    if ( ((step+1) % m_freq != 0) || warpx.get_load_balance_int() < 1 ) { return; }
+
+    // get number of boxes over all levels
+    auto nLevels = warpx.finestLevel() + 1;
+    int nBoxes = 0;
+    for (int lev = 0; lev < nLevels; ++lev)
+    {
+        const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(lev);
+        nBoxes += cost_heuristic->size();
+    }
+
+    // keep track of the max number of boxes, this is needed later on to fill
+    // the jagged array (in case each step does not have the same number of boxes)
+    m_nBoxesMax = std::max(m_nBoxesMax, nBoxes);
+
+    // resize and clear data array
+    m_data.resize(m_nDataFields*nBoxes, 0.0);
+    m_data.assign(m_nDataFields*nBoxes, 0.0);
+
+    // costs is recomputed, whether or not a load balance step; if not a load
+    // balance step, the costs vectors are empty; if it is a load balance step,
+    // remake level clears the costs vectors;
+    if (warpx.load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Heuristic)
+    {
+        warpx.ComputeCostsHeuristic();
+    }
```

Comment:
This will be problematic for the timers; during load balance when the level is remade, costs are reset to zero, but reduced diagnostics is called after load balance, meaning that costs have already been reset to zero when we want to output the reduced costs diagnostic.  We can recompute this for `costs_heuristic`, but we could not go back and re-time code blocks.  What is simplest here?  Do we want to output the costs reduced diagnostic before load balancing?  Or drag around copies of the load balance data, to be used only for reduced diagnostics, which we clear only after writing the cost reduced diagnostic?  Something else?  I'd prefer the first option if we can separate only output of costs from other reduced diagnostics.

Reply:
I think the main problem is that a diagnostic should not change the state of class WarpX **at all**. IMHO this must be respected in all diagnostics. Ideally, WarpX would be `const` in some way here, but we have not implemented a clean way to access all WarpX member variables in a const manner, so we have to rely on the developer's attention.

In this case, I would suggest to modify `WarpX::ComputeCostsHeuristic` so that it computes and returns the cost. When called in LoadBalancedHeuristic, the new costs would be stored in `WarpX::costs_heuristic`. Here, you would do whatever you want with the return value.

Now, as long as we give enough flexibility, it is the user responsibility to use them properly. Typically, 
```
load_balance_int = 10
my_costs.frequency = 5::10 # every 10 iterations, starting at 5 and until the end of the run
```
But we can keep this for later.